### PR TITLE
K8SPXC-609: update haproxy service only when changed

### PR DESCRIPTION
**Pull request by qJkee**
[K8SPXC-609](https://jira.percona.com/browse/K8SPXC-609) [&#10088;?&#10089;](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

Currently we are updating HAProxy service each reconcile loop, which cause issues with LoadBalancer.
This PR is fixing this issue by checking if HAProxy configuration changed

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| haproxy | passed |
| recreate | passed |
| storage | passed |
| scaling | passed |
| upgrade-haproxy | passed |
| init-deploy | passed |
| self-healing | failed |
| scaling-proxysql | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| limits | passed |
| monitoring-2-0 | passed |
| restore-to-encrypted-cluster | passed |
| smart-update | passed |
| affinity | passed |
| one-pod | passed |
| self-healing-advanced | passed |
| demand-backup | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| operator-self-healing | passed |
| users | passed |
| pitr | passed |
| upgrade-consistency | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| scheduled-backup | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | passed |
| scaling | passed |
| storage | passed |
| haproxy | passed |
| big-data | passed |
| recreate | passed |
| self-healing | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| security-context | passed |
| limits | passed |
| restore-to-encrypted-cluster | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| upgrade-proxysql | passed |
| affinity | passed |
| demand-backup | failed |
| operator-self-healing | passed |
| one-pod | passed |
| smart-update | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| pitr | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| demand-backup-encrypted-with-tls | failed |

File: pkg/pxc/service.go [hunk 1]
```diff
@@ -281,6 +296,13 @@ func NewServiceHAProxy(cr *api.PerconaXtraDBCluster) *corev1.Service {
 	return obj
 }
 
+func NewServiceHaproxyName(cr *api.PerconaXtraDBCluster) types.NamespacedName {
```
Comment (blez) [inline pkg/pxc/service.go:299]: Conciser rename to HaproxyServiceNamespacedName

File: pkg/pxc/service.go [hunk 2]
```diff
@@ -338,3 +360,10 @@ func NewServiceHAProxyReplicas(cr *api.PerconaXtraDBCluster) *corev1.Service {
 
 	return obj
 }
+
+func NewServiceHAProxyReplicasName(cr *api.PerconaXtraDBCluster) types.NamespacedName {
```
Comment (blez) [inline pkg/pxc/service.go:364]: Consider rename to HAProxyReplicasNamespacedName

File: pkg/controller/pxc/controller.go [hunk 3]
```diff
@@ -1109,3 +1099,40 @@ func createOrUpdateConfigmap(cl client.Client, configMap *corev1.ConfigMap) erro
...
+func updateServiceIfNeeded(cl client.Client, old, new *corev1.Service) error {
+	if !isServicesSpecEqual(&old.Spec, &new.Spec) || !reflect.DeepEqual(old.Annotations, new.Annotations) ||
+		!reflect.DeepEqual(old.Labels, new.Labels) || !reflect.DeepEqual(new.Spec.Selector, old.Spec.Selector) {
+		log.Info("UPDATING HAPROXY", "name", new.Name)
+		return cl.Update(context.TODO(), new)
+	}
+	return nil
+}
+
+func isServicesSpecEqual(old, new *corev1.ServiceSpec) bool {
```
Comment (blez) [inline pkg/controller/pxc/controller.go:1112]: Why pointers?

File: pkg/controller/pxc/controller.go [hunk 4]
```diff
@@ -1109,3 +1099,40 @@ func createOrUpdateConfigmap(cl client.Client, configMap *corev1.ConfigMap) erro
 
 	return nil
 }
+
+func updateServiceIfNeeded(cl client.Client, old, new *corev1.Service) error {
+	if !isServicesSpecEqual(&old.Spec, &new.Spec) || !reflect.DeepEqual(old.Annotations, new.Annotations) ||
+		!reflect.DeepEqual(old.Labels, new.Labels) || !reflect.DeepEqual(new.Spec.Selector, old.Spec.Selector) {
+		log.Info("UPDATING HAPROXY", "name", new.Name)
```
Comment (blez) [inline pkg/controller/pxc/controller.go:1106]: This log message looks aggressive :smile:

File: pkg/controller/pxc/controller.go [hunk 5]
```diff
@@ -343,115 +343,108 @@ func (r *ReconcilePerconaXtraDBCluster) Reconcile(request reconcile.Request) (re
...
-			//Upgrading service only if something is changed
-			if currentServiceReplicas.Spec.Type != o.Spec.HAProxy.ReplicasServiceType {
-				currentServiceReplicas.Spec.Ports = replicaPorts
-				currentServiceReplicas.Spec.Type = o.Spec.HAProxy.ReplicasServiceType
-			}
-			//Checking default ServiceType
-		} else if currentServiceReplicas.Spec.Type != corev1.ServiceTypeClusterIP {
-			currentServiceReplicas.Spec.Ports = replicaPorts
-			currentServiceReplicas.Spec.Type = corev1.ServiceTypeClusterIP
+			newHAProxyServiceReplicas.Spec.Type = o.Spec.PXC.ReplicasServiceType
```
Comment (blez) [inline pkg/controller/pxc/controller.go:427]: Is it really should be PXC ReplicasServiceType, not HAProxy?

File: pkg/pxc/service.go [hunk 6]
```diff
@@ -207,6 +215,13 @@ func NewServiceProxySQL(cr *api.PerconaXtraDBCluster) *corev1.Service {
 	return obj
 }
 
+func NewServiceProxySQLName(cr *api.PerconaXtraDBCluster) types.NamespacedName {
```
Comment (blez) [inline pkg/pxc/service.go:218]: Consider rename to ProxySQLServiceNamespacedName

File: pkg/pxc/service.go [hunk 7]
```diff
@@ -144,6 +145,13 @@ func NewServiceProxySQLUnready(cr *api.PerconaXtraDBCluster) *corev1.Service {
 	return obj
 }
 
+func NewServiceProxySQLUnreadyName(cr *api.PerconaXtraDBCluster) types.NamespacedName {
```
Comment (blez) [inline pkg/pxc/service.go:148]: Consider rename to ProxySQLUnreadyServiceNamespacedName

File: pkg/controller/pxc/controller.go [hunk 8]
```diff
@@ -1109,3 +1099,40 @@ func createOrUpdateConfigmap(cl client.Client, configMap *corev1.ConfigMap) erro
 
 	return nil
 }
+
+func updateServiceIfNeeded(cl client.Client, old, new *corev1.Service) error {
```
Comment (blez) [inline pkg/controller/pxc/controller.go:1103]: I think this func should not update service. It just should return true of false depend on is services are equal or not.

File: pkg/controller/pxc/controller.go [hunk 9]
```diff
@@ -1109,3 +1099,40 @@ func createOrUpdateConfigmap(cl client.Client, configMap *corev1.ConfigMap) erro
...
+		log.Info("UPDATING HAPROXY", "name", new.Name)
+		return cl.Update(context.TODO(), new)
+	}
+	return nil
+}
+
+func isServicesSpecEqual(old, new *corev1.ServiceSpec) bool {
+	if old.Type != new.Type || !isServicesPortsEqual(old.Ports, new.Ports) ||
+		old.ExternalTrafficPolicy != new.ExternalTrafficPolicy ||
+		!reflect.DeepEqual(old.LoadBalancerSourceRanges, new.LoadBalancerSourceRanges) {
```
Comment (blez) [inline pkg/controller/pxc/controller.go:1115]: It's easy co compare 2 slices without using reflection

File: pkg/controller/pxc/controller.go [hunk 10]
```diff
@@ -1109,3 +1099,40 @@ func createOrUpdateConfigmap(cl client.Client, configMap *corev1.ConfigMap) erro
 
 	return nil
 }
+
+func updateServiceIfNeeded(cl client.Client, old, new *corev1.Service) error {
+	if !isServicesSpecEqual(&old.Spec, &new.Spec) || !reflect.DeepEqual(old.Annotations, new.Annotations) ||
+		!reflect.DeepEqual(old.Labels, new.Labels) || !reflect.DeepEqual(new.Spec.Selector, old.Spec.Selector) {
```
Comment (blez) [inline pkg/controller/pxc/controller.go:1105]: We do not need to use reflection to compare maps.

File: pkg/controller/pxc/controller.go (hunk 9 above)
Comment (qJkee) [reply pkg/controller/pxc/controller.go:1115]: unfortunately, there are other fields set by k8s which we need to ignore)
That's why i have to do it manually

File: pkg/controller/pxc/controller.go (hunk 8 above)
Comment (qJkee) [reply pkg/controller/pxc/controller.go]: 068a20ad

File: pkg/pxc/service.go (hunk 7 above)
Comment (qJkee) [reply pkg/pxc/service.go]: fbb29694

File: pkg/pxc/service.go (hunk 6 above)
Comment (qJkee) [reply pkg/pxc/service.go]: fbb29694

File: pkg/controller/pxc/controller.go (hunk 5 above)
Comment (qJkee) [reply pkg/controller/pxc/controller.go]: 56e4fe9c

File: pkg/controller/pxc/controller.go (hunk 4 above)
Comment (qJkee) [reply pkg/controller/pxc/controller.go]: removed there)
068a20ad

File: pkg/controller/pxc/controller.go (hunk 3 above)
Comment (qJkee) [reply pkg/controller/pxc/controller.go]: 068a20ad

File: pkg/pxc/service.go (hunk 2 above)
Comment (qJkee) [reply pkg/pxc/service.go]: fbb29694

File: pkg/pxc/service.go (hunk 1 above)
Comment (qJkee) [reply pkg/pxc/service.go]: fbb29694

File: pkg/controller/pxc/controller.go (hunk 10 above)
Comment (qJkee) [reply pkg/controller/pxc/controller.go:1108]: DeepEqual looks more optimal there)

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| haproxy | passed |
| upgrade-haproxy | passed |
| recreate | passed |
| big-data | passed |
| scaling | passed |
| init-deploy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | failed |
| smart-update | passed |
| demand-backup | failed |
| operator-self-healing | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| upgrade-consistency | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| pitr | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| init-deploy | passed |
| upgrade-haproxy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| recreate | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| big-data | passed |
| scaling | passed |
| scaling-proxysql | passed |
| security-context | passed |
| storage | passed |
| self-healing | passed |
| self-healing-advanced | passed |
| operator-self-healing | passed |
| upgrade-proxysql | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| pitr | passed |
| demand-backup-encrypted-with-tls | passed |
| scheduled-backup | failed |

**Comment by JNKPercona** [comment]
PXC operator docker - perconalab/percona-xtradb-cluster-operator:PR-749-068a20ad

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| scaling | passed |
| haproxy | passed |
| restore-to-encrypted-cluster | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| demand-backup | passed |
| limits | passed |
| storage | passed |
| security-context | passed |
| pitr | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| demand-backup-encrypted-with-tls | passed |
| affinity | passed |
| self-healing-advanced | passed |
| scheduled-backup | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| upgrade-haproxy | passed |
| proxysql-sidecar-res-limits | passed |
| big-data | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |

File: pkg/controller/pxc/controller.go (hunk 9 above)
Comment (blez) [reply pkg/controller/pxc/controller.go:1118]: I ment we can sort them and compare index by index

File: pkg/controller/pxc/controller.go [hunk 11]
```diff
@@ -1109,3 +1102,40 @@ func createOrUpdateConfigmap(cl client.Client, configMap *corev1.ConfigMap) erro
...
+		for _, newSpec := range new {
+			if oldSpec.Name == newSpec.Name {
+				if oldSpec.Port != newSpec.Port || oldSpec.Protocol != newSpec.Protocol ||
+					oldSpec.TargetPort.IntVal != newSpec.TargetPort.IntVal {
+					return false
+				}
+			}
+		}
+	}
+	return true
```
Comment (blez) [inline pkg/controller/pxc/controller.go:1140]: It's easier to sort it by name and compare index by index.
Wnat if we have:
[name1,name2] and [name2,name3] ? 
lens are equal but names are not.
your algorithm will return true. But as I understand it should return false

Comment (blez) [reply pkg/controller/pxc/controller.go:1140]: This is actually a good example of func we can write test for

Comment (qJkee) [reply pkg/controller/pxc/controller.go]: nice catch
5f4fc8dd

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| recreate | passed |
| upgrade-haproxy | passed |
| scaling | passed |
| haproxy | passed |
| big-data | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| security-context | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| smart-update | passed |
| demand-backup | passed |
| affinity | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| pitr | passed |
| validation-hook | passed |
| demand-backup-encrypted-with-tls | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| init-deploy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| upgrade-haproxy | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| big-data | passed |
| tls-issue-cert-manager-ref | passed |
| scaling | passed |
| recreate | passed |
| validation-hook | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| security-context | passed |
| storage | passed |
| pitr | passed |
| self-healing | passed |
| demand-backup-encrypted-with-tls | passed |
| self-healing-advanced | passed |
| operator-self-healing | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| scheduled-backup | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| upgrade-haproxy | passed |
| recreate | passed |
| haproxy | passed |
| storage | passed |
| scaling | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| upgrade-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| smart-update | passed |
| demand-backup | passed |
| affinity | passed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| pitr | passed |
| users | passed |
| upgrade-consistency | passed |
| demand-backup-encrypted-with-tls | passed |
| scheduled-backup | failed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

**Review by blez — DISMISSED** [review]

**Pull request closed without merging**